File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	static int[] nums, oper, visited;
	static int N;
	static int min, max;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		nums = new int[N];
		oper = new int[N];
		visited = new int[4];
		min = 1000000000;
		max = -1000000000;

		StringTokenizer st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			nums[i] = Integer.parseInt(st.nextToken());
		}
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < 4; i++) {
			visited[i] = Integer.parseInt(st.nextToken());
		}

		perm(1);
		System.out.println((int) max);
		System.out.println((int) min);

	}

	private static void perm(int depth) {
		if (depth == N) {
			int c = calc();
			min = min > c ? c : min;
			max = max < c ? c : max;
			return;
		}
		for (int i = 0; i < 4; i++) {
			if(visited[i]==0) continue;
			visited[i]--;
			oper[depth] = i;
			perm(depth+1);
			visited[i]++;
		}

	}

	private static int calc() {
		int result = nums[0];
		for (int i = 1; i < N; i++) {
			if (oper[i] == 0) {
				result += nums[i];
			} else if (oper[i] == 1) {
				result -= nums[i];
			} else if (oper[i] == 2) {
				result *= nums[i];
			} else if (oper[i] == 3) {
				result /= nums[i];
			}
		}
		return result;
	}

}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
# 시작시간 3시 38분
# 제출시간 3시 56분
# 소요시간 18분 + 1분
# 시도횟수 2회(max초기값 처리)
# 나눗셈 => 음수일 경우 -(-num1 / num2)


# 입력 완료 로직 시작
def perm(depth):
    if depth == N:  # 다 골랐으면
        cc = select_clac()  # select_opers순서대로 계산
        ans[0] = max(ans[0], cc)  # mx갱신
        ans[1] = min(ans[1], cc)  # mn갱신
        return
    for i in range(4):
        if opers[i] == 0: # 남은거 없으면 패스
            continue
        select_opers[depth] = i
        opers[i] -= 1  # 사용했으니 하나 줄이기
        perm(depth + 1)  # 재귀 들어갔다 나오면
        opers[i] += 1  # 원상복구


# num와 select_oper을 연산한 값 리턴
def select_clac():
    result = nums[0]
    for i in range(1, N):
        result = clac(result, nums[i], select_opers[i])
    return result


# 사칙연산 결과값 리턴
def clac(num1, num2, op):
    if op == 0:
        return num1 + num2
    elif op == 1:
        return num1 - num2
    elif op == 2:
        return num1 * num2
    elif op == 3:  # 음수 나눗셈 처리
        return num1 // num2 if num1 >= 0 else -(-num1 // num2)


N = int(input())
nums = list(map(int, input().split()))
opers = list(map(int, input().split()))  # 4개 / visited처럼 사용
select_opers = [0] * N  # 1부터 계산식 넣음. 즉, 계산은 num[i] oper[i+1] num[i+1]
ans = [-20000000000, 20000000000]  # 최대, 최소

perm(1)
print(*ans, sep="\n")